apps/designer/app/designer/shared/assets/use-assets.ts

```diff
@@ -1,56 +1,89 @@
-import { useFetcher } from "@remix-run/react";
+import { useEffect, useMemo } from "react";
 import {
   AssetType,
   filterByType,
   MAX_UPLOAD_SIZE,
   toBytes,
+  type Asset,
 } from "@webstudio-is/asset-uploader";
 import { toast } from "@webstudio-is/design-system";
 import ObjectID from "bson-objectid";
-import { useMemo } from "react";
 import { restAssetsPath } from "~/shared/router-utils";
 import { useAssets as useAssetsState, useProject } from "../nano-states";
 import { sanitizeS3Key } from "@webstudio-is/asset-uploader";
 import { PreviewAsset } from "./types";
+import { usePersistentFetcher } from "~/shared/fetcher";
+import type { ActionData } from "~/designer/shared/assets";
+import {
+  FetcherData,
+  normalizeErrors,
+  toastUnknownFieldErrors,
+} from "~/shared/form-utils";
+import { Publish } from "~/shared/pubsub";
+import { useFetcher } from "@remix-run/react";
 
-const toPreviewAssets = (formData: FormData): Promise<PreviewAsset[]> => {
-  const assets: Array<Promise<PreviewAsset>> = [];
+declare module "~/shared/pubsub" {
+  export interface PubsubMap {
+    updateAssets: Array<Asset>;
+  }
+}
 
-  for (const entry of formData) {
-    const file = entry[1] as File;
-    const promise: Promise<PreviewAsset> = new Promise((resolve, reject) => {
-      const reader = new FileReader();
-      reader.addEventListener("load", (event) => {
-        const dataUri = event?.target?.result;
-        if (dataUri === undefined) {
-          return reject(new Error(`Could not read file "${file.name}"`));
-        }
+export const usePublishAssets = (publish: Publish) => {
+  const [assets] = useAssetsState();
+  useEffect(() => {
+    publish({
+      type: "updateAssets",
+      payload: assets.filter(
+        (asset) => asset.status === "uploaded"
+      ) as Array<Asset>, // TS doesn't understand we filtered out PrevewAssets
+    });
+  }, [assets, publish]);
+};
+
+export type UploadData = FetcherData<ActionData>;
+
+const toPreviewAssets = (
+  formsData: Array<FormData>
+): Promise<PreviewAsset[]> => {
+  const assets: Array<Promise<PreviewAsset>> = [];
+  for (const formData of formsData) {
+    for (const entry of formData) {
+      const file = entry[1] as File;
```

Comment:
as I remember `instanceof` operator should work with file
```js
var a= new File([23], '2323')
if (a instanceof File) {}
```

Reply:
got it